sql/driver/conn.go

```diff
@@ -376,3 +522,225 @@ func (c *conn) lookupDatabase(name string) (uint32, error) {
 	}
 	return uint32(gr.ValueInt()), nil
 }
+
+func (c *conn) processColumns(desc *structured.TableDescriptor,
+	node parser.Columns) ([]structured.ColumnDescriptor, error) {
+	if node == nil {
+		return desc.Columns, nil
+	}
+
+	cols := make([]structured.ColumnDescriptor, len(node))
+	for i, n := range node {
+		switch nt := n.(type) {
+		case *parser.StarExpr:
+			return c.processColumns(desc, nil)
+		case *parser.NonStarExpr:
+			switch et := nt.Expr.(type) {
+			case *parser.ColName:
+				// TODO(pmattis): If et.Qualifier is not empty, verify it matches the
+				// table name.
+				var err error
+				cols[i], err = findColumnByName(desc, et.Name)
+				if err != nil {
+					return nil, err
+				}
+			default:
+				return nil, fmt.Errorf("unexpected node: %T", nt.Expr)
+			}
+		}
+	}
+
+	return cols, nil
+}
+
+func (c *conn) processInsertRows(node parser.InsertRows) (*rows, error) {
+	switch nt := node.(type) {
+	case parser.Values:
+		r := &rows{pos: -1}
+		for _, row := range nt {
+			switch rt := row.(type) {
+			case parser.ValTuple:
+				var vals []driver.Value
+				for _, val := range rt {
+					switch vt := val.(type) {
+					case parser.StrVal:
+						vals = append(vals, string(vt))
+					case parser.NumVal:
+						vals = append(vals, string(vt))
+					case parser.ValArg:
+						return nil, fmt.Errorf("TODO(pmattis): unsupported node: %T", val)
+					case parser.BytesVal:
+						vals = append(vals, string(vt))
+					default:
+						return nil, fmt.Errorf("TODO(pmattis): unsupported node: %T", val)
+					}
+				}
+				r.rows = append(r.rows, vals)
+			case *parser.Subquery:
+				return nil, fmt.Errorf("TODO(pmattis): unsupported node: %T", row)
+			}
+		}
+		return r, nil
+	case *parser.Select:
+		return c.query(nt, nil)
+	case *parser.Union:
+		return c.query(nt, nil)
+	}
+	return nil, fmt.Errorf("TODO(pmattis): unsupported node: %T", node)
+}
+
+func findColumnByName(desc *structured.TableDescriptor, name string) (
+	structured.ColumnDescriptor, error) {
+	// TODO(pmattis): This should really be done using a map from column name to
+	// column descriptor.
+	for _, c := range desc.Columns {
+		if c.Name == name {
+			return c, nil
+		}
+	}
+	c := structured.ColumnDescriptor{}
+	return c, fmt.Errorf("column \"%s\" does not exist", name)
+}
+
+func findColumnByID(desc *structured.TableDescriptor, id uint32) (
+	structured.ColumnDescriptor, error) {
+	// TODO(pmattis): This should really be done using a map from column id to
+	// column descriptor.
+	for _, c := range desc.Columns {
+		if c.ID == id {
+			return c, nil
+		}
+	}
+	c := structured.ColumnDescriptor{}
+	return c, fmt.Errorf("column-id \"%d\" does not exist", id)
+}
+
+func encodeIndexKeyPrefix(tableID, indexID uint32) []byte {
```

Comment:
Yes, or in `structured`, though I'm hesitant to move them right now when they depend on `driver.Value`. I'll add a TODO that this is not a permanent location.